flang/lib/Lower/OpenMP.cpp

```diff
@@ -426,14 +425,10 @@ void DataSharingProcessor::collectSymbols(
                              /*collectSymbols=*/true,
                              /*collectHostAssociatedSymbols=*/true);
   for (Fortran::lower::pft::Evaluation &e : eval.getNestedEvaluations()) {
-    if (e.hasNestedEvaluations())
+    if (e.hasNestedEvaluations() && !e.isConstruct())
```

Comment:
I had a hard time trying to understand how `collectSymbols()` work.
This is how I understand it:
- First, all symbols used in `eval` that match `flag` are collected into `defaultSymbols`, including host associated symbols
   (why host associated symbols need to be collected? To look at the outer scope? But if the symbol in the outer scope  matches the default DSA, does anything need to be done?)
- For each evaluation that is a direct child of `eval`:
  - If the evaluation has children and is not a Fortran construct (such as if, do, etc)
    - Add the symbols used in it, that match `flag`, to `symbolsInNestedRegions`, skipping host associated symbols
      (because otherwise symbols from the top `eval` level could also be collected?)
- Later, in `defaultPrivatize`, `symbolsInNestedRegions` are excluded from `defaultSymbols`, presumably because each nested region will handle them appropriately. Then, the remaining symbols are privatized, with some exceptions, such as procedures.

Is my understanding correct?

It would be nice to add a brief comment to `collectSymbols`, explaining what it does.

Reply:
Thanks @luporl for the review. Yes, the understanding is correct. Since each nested region needs to handle the symbols separately, we needed to make `collectSymbols` as it is now. In fact, in response to Kiran's comment above, I have given the lowered IR for combinations of nested constructs with default clause. You could check if you are ok with the current behaviour.

And sure, I'll capture more details on `collectSymbols` with this PR.

Reply:
Thanks for the explanation @NimishMishra. The current behavior seems ok to me.
I'm planning to use `collectSymbols` to handle symbols with implicit DSA, such as those of `task`, that must be `firstprivate` if no `default` clause is present (and outer context doesn't define it as shared).